e2e/nomostest/reset.go

```
@@ -0,0 +1,513 @@
+			}); err != nil {
+				return err
+			}
+		}
+
+		// Print reconciler logs in case of failure.
+		// This ensures the logs are printed, even if the reconciler is deleted.
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		go TailReconcilerLogs(ctx, nt, RootReconcilerObjectKey(rsNN.Name))
+
+		// DeletePropagationBackground is required when deleting RSyncs with
+		// dependencies that have owners references. Otherwise the reconciler
+		// and dependenencies will be garbage collected before the finalizer
+		// can delete the managed resources.
+		// TODO: Remove explicit Background policy after the reconciler-manager finalizer is added.
+		nt.T.Logf("[RESET] Deleting RootSync %s", rsNN)
+		if err := nt.Delete(rs, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+			return err
+		}
+	}
+	tg := taskgroup.New()
+	for _, item := range rsList {
+		rs := &item
+		rsNN := client.ObjectKeyFromObject(rs)
+		nt.T.Logf("[RESET] Waiting for deletion of RootSync %s ...", rsNN)
+		tg.Go(func() error {
+			return WatchForNotFound(nt, kinds.RootSyncV1Beta1(), rs.Name, rs.Namespace)
+		})
+	}
+	return tg.Wait()
+}
+
+// ResetRepoSyncs cleans up one or more RepoSyncs and all their managed objects.
+// Use this for cleaning up RepoSyncs in tests that use delegated control.
+//
+// To ensure the reconcile finalizer has permission to delete managed resources,
+// ClusterRole and RoleBindings will be created and then later deleted.
+// This also cleans up any CRs, RBs, and CRBs left behind by delegated control.
+func ResetRepoSyncs(nt *NT, rsList []v1beta1.RepoSync) error {
+	if len(rsList) == 0 {
+		// Clean up after `setupDelegatedControl`
+		return deleteRepoSyncClusterRole(nt)
+	}
+
+	nt.T.Logf("[RESET] Deleting RepoSyncs (%d)", len(rsList))
+
+	record := newResetRecord()
+
+	// Apply ClusterRole with the permissions specified by this test.
+	rsCR := nt.RepoSyncClusterRole()
+	if err := nt.Apply(rsCR); err != nil {
```

ditto here for your thoughts to `defer` deleting the `ClusterRole`

---

We could put `deleteRepoSyncClusterRole` in a defer. Tho I'm skeptical about trying to clean up after failure in the clean up function... I kinda prefer cleanup that fails hard and stops, so you know where it broke and can debug without the state changing afterwards.

---

Yeah that seems reasonable, I don't have any real opposition here just seemed like a viable alternative